### PR TITLE
Add A.WithContext

### PR DESCRIPTION
**Pull request by maniak89**
## Why

There was a need to pass additional parameters between actions without using environment variables because the tasks are executed in parallel. No other simpler mechanism was provided.

## What

Add `A.WithContext` that creates a derived `A` with a changed context.
Thanks to it `A` can be reused to pass cancelation and values via context to the helper functions.

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated.
- [x] The code changes follow [Effective Go](https://golang.org/doc/effective_go).
- [x] The code changes follow [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments).
- [x] The code changes are covered by tests.

### CODE REVIEW

**Comment by pellared** [comment]
Can you please provide an example how would the new method be used?

**Comment by pellared** [comment]
I do not understand how this would help. How changing the Context of the task that is being executed would help?

Why not taking advantage of something provided by Go itself, like `sync/atomic`, `sync.Mutex`, `sync.Cond`.

I think this can be somehow related to:
- https://github.com/goyek/goyek/issues/435

**Comment by maniak89** [comment]
>Can you please provide an example how would the new method be used?
>

`

type ctxKey struct{}

var ctxVal ctxKey

func someTask(flow *goyek.Flow) *goyek.DefinedTask {
	task := flow.Define(goyek.Task{
		Name: "simple use from ctx",
		Action: func(a *goyek.A) {
			args,ok := a.Context().Value(ctxVal).(map[string]string) // extract from replaced context
			if !ok {
				return
			}

			fmt.Printf("%#v", args)
		},
	})

	addSubTasks(flow, task, []map[string]string{
		{
			"some":"111",
		},
		{
			"some":"2222",
		},
	})

	return task
}

func addSubTasks(flow *goyek.Flow, src *goyek.DefinedTask, subArgs []map[string]string) []*goyek.DefinedTask {
	if len(src.Deps()) == 0 && src.Action() == nil {
		return nil
	}

	result := make([]*goyek.DefinedTask, 0, len(subArgs))
	for index, subArg := range subArgs {
		result = append(result, flow.Define(goyek.Task{
			Name: src.Name() + "_" + strconv.Itoa(index), // for simple only name
			Action: func(a *goyek.A) {
				src.Action()(a.WithContext(context.WithValue(a.Context(), ctxVal, subArg))) // put to context
			},
		}))
	}

	return result
}

`

> I do not understand how this would help. How changing the Context of the task that is being executed would help?
> 
> Why not taking advantage of something provided by Go itself like `sync/atomic`, `sync.Mutex`, `sync.Cond`?
> 
> I think this can be somehow related to:
> 
> * [Add pools support #435](https://github.com/goyek/goyek/issues/435)

This is not relevant to this problem. I am trying to implement "arguments" for a task via context.

**Comment by pellared** [comment]
You do not need anything added. You can refactor your `addSubTasks` to not accept `src *goyek.DefinedTask`  but some functions.

Just a prototype

```go
type ctxKey struct{}

var ctxVal ctxKey

func foo(ctx context.Context, _ *goyek.A) {
			args,ok := ctx.Value(ctxVal).(map[string]string) // extract from replaced context
			if !ok {
				return
			}

			fmt.Printf("%#v", args)
}

func someTask(flow *goyek.Flow) *goyek.DefinedTask {
	task := flow.Define(goyek.Task{
		Name: "simple use from ctx",
		Action: func(a *goyek.A) {
			foo(a.Context())
		},
	})

	addSubTasks(a.Context(), a.Name(), flow, task, []map[string]string{
		{
			"some":"111",
		},
		{
			"some":"2222",
		},
	})

	return task
}

func addSubTasks(ctx context.Context, flow *goyek.Flow, name string, src func(context.Context, *goyek.A) , subArgs []map[string]string) []*goyek.DefinedTask {
	if len(src.Deps()) == 0 && src.Action() == nil {
		return nil
	}

	result := make([]*goyek.DefinedTask, 0, len(subArgs))
	for index, subArg := range subArgs {
		result = append(result, flow.Define(goyek.Task{
			Name: name + "_" + strconv.Itoa(index), // for sample only name
			Action: func(a *goyek.A) {
				src(context.WithValue(ctx, ctxVal, subArg), a) // put to context
			},
			Parallel: true,
		}))
	}

	return result
}
```

**Comment by pellared** [comment]
Getting an action like this `src.Action()` and passing the arguments seems like a code smell.

Moreover,  I am pretty sure that mutating the context during the execution can lead to data races. Especially that you are running parallel tasks. Context should be "passed down" and not being mutated. 

Notice that even [`http.Request.WithContext](https://cs.opensource.google/go/go/+/refs/tags/go1.24.0:src/net/http/request.go;l=368) makes a copy of the request. However, `goyek.A` is not just a "request" which is similar to "context". It also acts as "response writer". You should rather see it more like `testing.T` (which has `t.Context()` since Go 1.24 :wink: ).

**Comment by maniak89** [comment]
Of course, it is possible in case of simple calls. In my example, the code is much simplified. I use DefinedTask, because the task can have dependencies, and I need to wrap them too. As an example - the task to collect depends on the task to generate something, copy or something like that. And this task should be executed differently for different platforms - there can be different paths, variable values ​​- all this is simply passed to map[string]string and passed along the entire dependency tree. Using setEnv is impossible with Parallel: true because the values ​​are erased by other tasks.

> Moreover, I am almost sure that changing the context during execution can lead to data races. Especially considering that you are running parallel tasks. The context must be "passed" and not changed.

Parallelization occurs at the moment of Action occurrence from Input, it cannot be executed in parallel. Each action exists at a time which is no longer parallelized and one specific action is executed at one point in time. Changing the ctn is safe from this point of view.

You cannot copy because otherwise the cleanup functions will not work (if you ask to execute cleanup during the action executing, the slice can be (and will probably be) changed and accordingly the one from which the copy was taken will not receive this cleanup function.

> (which has t.Context() since Go 1.24 😉 )
Unfortunately, according to the requirements, I cannot use higher than 1.22

**Comment by pellared** [comment]
@maniak89, what is the issue with the pattern I proposed here: https://github.com/goyek/goyek/pull/453#issuecomment-2671465912?

**Comment by pellared** [comment]
>  I use DefinedTask, because the task can have dependencies, and I need to wrap them too. 

I think you would need to bring this to the example as I think you could simply wrap functions as I showed instead of calling actions taken from other tasks.

**Comment by pellared** [comment]
> You cannot copy because otherwise the cleanup functions will not work (if you ask to execute cleanup during the action executing, the slice can be (and will probably be) changed and accordingly the one from which the copy was taken will not receive this cleanup function.

Gotcha

The issue of adding the ability to mutate the context of A is that called function can change the result of a.Context() (called by the.l caller). I do not find such side effect as acceptable.

**Comment by maniak89** [comment]
There is an idea - to have a copy and at the same time not break the cleaning, wait a moment

**Comment by maniak89** [comment]
@pellared, please take a look

**Comment by pellared** [comment]
> have a copy and at the same time not break the cleaning, wait a moment

That sounds good to me and I agree this can be very useful. Please also add some tests and consider edge cases to avoid regression bugs.

File: a.go [hunk 1]
```diff
@@ -37,6 +37,49 @@ func (a *A) Context() context.Context {
...
+	defer a.mu.Unlock()
+
+	result := &A{
+		ctx:     ctx,
+		name:    a.name,
+		output:  a.output,
+		logger:  a.logger,
+		failed:  a.failed,
+		skipped: a.skipped,
+	}
```
Comment (pellared) [inline a.go:57]: The issues you encountered with cleanup would be the same for all other mutex protected fields such as `failed` and `skip`.

For instance calling `a.FailNow()` on the returned instance would not cause the original to be marked as failed.

Comment (pellared) [reply a.go:57]: Does the same mutex has to be used for `failed`, `skipped` and `cleanups`? If not then maybe we could change the type for `failed` and `skipped` to `*atomic.Bool`? I am not sure if using a pointer to atomic is even good/possible/safe.

File: a.go [hunk 2]
```diff
@@ -37,6 +37,49 @@ func (a *A) Context() context.Context {
...
+			}
+
+			result.mu.Unlock()
+
+			if cleanup == nil {
+				return
+			}
+
+			cleanup()
+		}
```
[the remark this answers is not in the document]
Comment (maniak89) [reply a.go]: Yes, this is copy-paste, I already refactored it while writing tests for the message above =)

File: a.go (hunk 1 above)
Comment (maniak89) [reply a.go]: It seems to me that this is not safe, so look at my solution - it seems to me that it is clearer and safer.

File: a.go [hunk 3]
```diff
@@ -95,9 +125,14 @@ func (a *A) Failed() bool {
 
 // Fail marks the function as having failed but continues execution.
 func (a *A) Fail() {
+	var call func()
 	a.mu.Lock()
 	a.failed = true
+	call = a.failedCall
```
Comment (pellared) [inline a.go:131]: I do not think this need to be inside the lock. `a.failedCall` does not look to be ever mutated.

Probably this will work.


```go
func (a *A) Fail() {
	if a.failedCall != nil {
		a.failedCall()
		return
	}
	a.mu.Lock()
	a.failed = true
	a.mu.Unlock()
}
```

File: a.go [hunk 4]
```diff
@@ -315,19 +355,27 @@ func (a *A) runCleanups(finished *bool, panicVal *interface{}, panicStack *[]byt
...
+	a.mu.Unlock()
+	if cleanup == nil {
+		return false
+	}
+	cleanup()
+	return true
+}
+
+func (a *A) callCleanups() {
+	for a.callLastCleanup() {
```
Comment (pellared) [inline a.go:379]: Please inline `callLastCleanup` :wink:

File: a.go [hunk 5]
```diff
@@ -315,19 +355,27 @@ func (a *A) runCleanups(finished *bool, panicVal *interface{}, panicStack *[]byt
...
+func (a *A) callLastCleanup() bool {
+	var cleanup func()
+	a.mu.Lock()
+	if len(a.cleanups) > 0 {
+		last := len(a.cleanups) - 1
+		cleanup = a.cleanups[last]
+		a.cleanups = a.cleanups[:last]
+	}
+	a.mu.Unlock()
+	if cleanup == nil {
```
Comment (pellared) [inline a.go:371]: I probably discovered an already existing bug (please not to fix it in this PR).
If an action calls `a.Cleanup(nil)` then will would prevent previously registered to not being called.
Am I correct? If so then we can fix it in a separate PR.

**Review by pellared — COMMENTED** [review]
Side note:
I was planning to refactor the codebase at least a little for months...
I am happy that it is not in a such bad state that makes it impossible to contribute for others.
Writing code alone without reviews is such a strange thing :sweat_smile:

File: a.go (hunk 3 above)
Comment (maniak89) [reply a.go:131]: I think we just need to change the call in places so that the flag is set on both the copy and the parent.

File: a.go (hunk 5 above)
Comment (pellared) [reply a.go]: I think this should be something more or less like

```go
	for {
		var cleanup func()
		a.mu.Lock()
		if len(a.cleanups) == 0 {
			a.mu.Unlock()
			cleanupFinished = true
			return
		}
		last := len(a.cleanups) - 1
		cleanup = a.cleanups[last]
		a.cleanups = a.cleanups[:last]
		a.mu.Unlock()
		cleanup()
	}
```

Comment (pellared) [reply a.go]: > I didn't try to fix this problem, and if I did, it wasn't on purpose :) 

No. You have not fixed or introduced a problem. It is an exploratory discovery during review :sweat_smile: 
I will make an issue out of my comment.

Comment (pellared) [reply a.go]: https://github.com/goyek/goyek/issues/454

**Review by pellared — COMMENTED** [review]
Can you please add a changelog entry?

Looks good. I very appreciate your hard work :heart: 

However, I need more time for debugging to make sure that we have not missed any corner cases. I will do my best to do it tomorrow.

File: a.go [hunk 6]
```diff
@@ -22,21 +22,51 @@ import (
...
+	cleanups    []func()
 }
 
 // Context returns the run context.
 func (a *A) Context() context.Context {
 	return a.ctx
 }
 
+// WithContext returns a shallow copy of a with its context changed
+// to ctx. The provided ctx must be non-nil.
```
[the remark this answers is not in the document]
Comment (maniak89) [reply a.go:43]: Actually, I used part of the comment from https://pkg.go.dev/net/http#Request.WithContext. It's not hard for me, I'll change it as you suggest =)

Of course, you as the author know the problems better, that's why I'm trying to make it so that it suits you and at the same time fulfills my task.

**Comment by maniak89** [comment]
> Side note:
> I was planning to refactor the codebase at least a little for months...
> I am happy that it is not in a such bad state that makes it impossible to contribute for others.
> Writing code alone without reviews is such a strange thing 😅

I understand you perfectly. In turn, I'm glad that you at least somehow respond to attempts to improve or make necessary changes for users. I've tried many times to do something good for the open source community, but, as a rule, everything came up against the "company policy" of the supposedly "open" but actually trying to make money, or they simply did not understand the cause-and-effect relationships, why I suddenly propose changes. I will not go into details, just an example - there is a library that works with some DB, and it can convert incoming values bin the required format, but it is impossible to install a custom date parser there. In my opinion, this is wrong, since it strictly specifies RFS3339, and dates can be in a different format =) From the description above, it is clear that I use my fork where this problem is solved, but the authors did not accept this approach.

File: a_test.go [hunk 7]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+					}
+
+					chkFunc(t, result, "child result "+strconv.Itoa(index)+" is false")
+				}
+			})
+		}
+	}
+}
+
+func TestA_WithContextFatal(t *testing.T) {
```
Comment (pellared) [inline a_test.go:283]: The test structure of `TestA_WithContext`, `TestA_WithContextFatal`, `TestA_WithContextSkipped` tests is very similar. I think we should extract the test cases type. See: https://google.github.io/styleguide/go/decisions#data-driven-test-cases

File: a_test.go [hunk 8]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+						Name: "child task " + strconv.Itoa(i),
+						Action: func(a *goyek.A) {
+							a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+							onceCall(t, a.Name())()
+
+							if throwPanic {
+								a.Cleanup(func() {
+									panic("some panic")
+								})
+							}
```
Comment (pellared) [inline a_test.go:245]: Conditional logic like this makes tests very hard to maintain. I think there are two ways to solve it:
- duplicating the logic (adding a new test function e.g. `TestA_WithContextPanic`)
- adding some `func()` test case parameters to have common code for all `TestA_WithContext`, `TestA_WithContextFatal`, `TestA_WithContextSkipped`

I think it would be good to try to add some test case parameters as function hooks.

File: a_test.go [hunk 9]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+				},
+			})
+
+			childTasks := make(goyek.Deps, 10)
+			childTasksResults := make([]bool, len(childTasks))
+
+			task := flow.Define(goyek.Task{
+				Name: "task",
+				Action: func(a *goyek.A) {
+					a.Cleanup(onceCall(t, a.Name()+" cleanup"))
```
Comment (pellared) [inline a_test.go:313]: I think we  add something like

```go
a.Cleanup(func() {
  failed := a.Failed()
  assertTrue(t, failed, "a.Failed() should return true for "+a.Name())
})
```

Same for skipped.

File: a_test.go [hunk 10]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+}
+
+func onceCall(t *testing.T, name string) func() {
+	t.Helper()
+
+	var callsCount int
+	var callsCountM sync.Mutex
+
+	t.Cleanup(func() {
+		assertEqual(t, callsCount, 1, "unexpected number of calls '"+name+"'")
```
Comment (pellared) [inline a_test.go:487]: It is hard to determine if reading `callsCount` is always thread safe here.

You should use `atomic.Int32` or add `callsCountM.Lock()` in the cleanup:

```suggestion
		callsCountM.Lock()
		defer callsCountM.Unlock()
		assertEqual(t, callsCount, 1, "unexpected number of calls '"+name+"'")
```

From https://go.dev/ref/mem

> Don't be clever. 

:wink:

File: a_test.go [hunk 11]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+					onceCall(t, a.Name())()
+
+					k, ok := a.Context().Value(ctxKey).(int)
+					if !ok {
+						return
+					}
+
+					childTasksResults[k] = true
+
+					a.Fatal()
```
Comment (pellared) [inline a_test.go:323]: We should call `onceCall` a few times to make sure that the derived `A` interrupts the action.
Otherwise we are just testing the `t.Cleanup` from `testing` package here as mentioned previously :laughing: 

Same comment applies to `a.Skip()`


```suggestion
					onceFn := onceCall(t, a.Name())
					onceFn()

					k, ok := a.Context().Value(ctxKey).(int)
					if !ok {
						return
					}

					childTasksResults[k] = true

					a.Fatal()
					onceFn()
```

File: a_test.go [hunk 12]
```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+
+				flow.SetOutput(io.Discard)
+				loggerSpy := &helperLoggerSpy{}
+				flow.SetLogger(loggerSpy)
+
+				depsTask := flow.Define(goyek.Task{
+					Name: "deps",
+					Action: func(a *goyek.A) {
+						a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+						onceCall(t, a.Name())()
```
Comment (pellared) [inline a_test.go:210]: This does not test anything apart from the `t.Cleanup` from `testing` package :wink: 
I do not think there is a possibility that this would ever fail (apart from a bug in `t.Cleanup`).

**Review by pellared — COMMENTED** [review]
Implementation-wise I think everything is good :medal_sports: 

Please let me know if you are OK with addressing the comments regarding the test code.
I would prefer doing the refactoring/improvements together rather than on my own.
At the same time, I do not want make your contribution a frustrating experience.

File: a_test.go (hunk 10 above)
Comment (pellared) [reply a_test.go:487]: PS. There is an edge case where it could race. When the function returned by `onceFunc` would be called in seperate goroutine. Then it can run concurrently with test cleanup thus produce a race condition.

File: a_test.go (hunk 8 above)
Comment (maniak89) [reply a_test.go:245]: They are similar in many ways, but also similar in places... I try to unite them with helpers.

File: a.go (hunk 1 above)
Comment (pellared) [reply a.go]: @maniak89, I just wanted to show that we could use pointers to atomic (it is safe): https://go.dev/play/p/4FmvNRZfrhB

And I see that `sync.Mutex` as a pointer is also used: https://stackoverflow.com/a/67786582

However, I do not want to go back and forth. I can always refactor it later. This is an implementation detail. I may want to make benchmarks to see what is more efficient in regular uses cases.

**My guess it that for most users  would not ever call `WithContext`. Your proposal should be more efficient in this scenario.**

File: a_test.go (hunk 8 above)
Comment (pellared) [reply a_test.go:245]: Consider time-boxing the effort. If uniting would be not straightforward, then I prefer more readable duplicated code. We can look into refactoring the duplication in a separate PR.

File: a_test.go [hunk 13]
```diff
@@ -183,42 +186,360 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+
+	depsTask := flow.Define(goyek.Task{
+		Name: "deps",
+		Action: func(a *goyek.A) {
+			a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+		},
+	})
+
+	childTasksResults := make([]bool, childCount)
+	return flow, flow.Define(goyek.Task{
```
Comment (pellared) [inline a_test.go:418]: Is there any reason to return `*goyek.DefinedTask` instead of the action `func(*goyek.A)`?
Does it have any relevance to `A.WithContext`?

File: a_test.go [hunk 14]
```diff
@@ -183,42 +186,360 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+	flow.SetOutput(io.Discard)
+	loggerSpy := &helperLoggerSpy{}
+	flow.SetLogger(loggerSpy)
+
+	depsTask := flow.Define(goyek.Task{
+		Name: "deps",
+		Action: func(a *goyek.A) {
+			a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+		},
+	})
```
Comment (pellared) [inline a_test.go:415]: Does it have any relevance to `A.WithContext`?

File: a_test.go [hunk 15]
```diff
@@ -183,42 +186,360 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+	t.Helper()
+
+	childTasks := make(goyek.Deps, count)
+	for i := 0; i < count; i++ {
+		i := i
+		childTasks[i] = flow.Define(goyek.Task{
+			Name:     "child task " + strconv.Itoa(i),
+			Action:   actionBuilder(i),
+			Deps:     goyek.Deps{src},
+			Parallel: parallel,
```
Comment (pellared) [inline a_test.go:446]: Does this functionality have anything to do with `Parallel` execution?
The `a` with modified context are passed via callstack.
I think we can remove all `parallel` tests as we are testing existing functionality of parallel invocation and not how `WithContext` behaves.

File: a_test.go [hunk 16]
```diff
@@ -183,42 +186,360 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
...
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			flow, task, childTasksResults, loggerSpy := prepareFlowAndTask(t, 3, ctxKey, func(a *goyek.A, i int) {
+				a.Fatal(i)
+
+				a.Cleanup(func() {
+					failed := a.Failed()
+					assertTrue(t, failed, "a.Failed() should return true for "+a.Name())
+				})
```
Comment (pellared) [inline a_test.go:284]: This will never be called. You need to register cleanup before. `a.Fatal`.

File: a.go (hunk 1 above)
Comment (pellared) [reply a.go]: Actually, I think there needs to be pointers.. Otherwise the "failed" and "skipped" will not get propagated to derived `a` which seems not what a user may expect.

**Review by pellared — CHANGES_REQUESTED** [review]
I have written I quite simple table driven test that in my opinion covers all the major corner cases:

```go
unc TestA_WithContext(t *testing.T) {
	testCases := []struct {
		desc        string
		fn          func(a, a2 *goyek.A)
		wantStatus  goyek.Status
		wantFailed  bool
		wantSkipped bool
	}{
		{
			desc:       "Pass",
			fn:         func(_, _ *goyek.A) {},
			wantStatus: goyek.StatusPassed,
		},
		{
			desc: "Fail",
			fn: func(a, _ *goyek.A) {
				a.FailNow()
			},
			wantStatus: goyek.StatusFailed,
			wantFailed: true,
		},
		{
			desc: "FailDerived",
			fn: func(_, a2 *goyek.A) {
				a2.FailNow()
			},
			wantStatus: goyek.StatusFailed,
			wantFailed: true,
		},
		{
			desc: "Skip",
			fn: func(a, _ *goyek.A) {
				a.SkipNow()
			},
			wantStatus:  goyek.StatusSkipped,
			wantSkipped: true,
		},
		{
			desc: "SkipDerived",
			fn: func(_, a2 *goyek.A) {
				a2.SkipNow()
			},
			wantStatus:  goyek.StatusSkipped,
			wantSkipped: true,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.desc, func(t *testing.T) {
			sb := &strings.Builder{}

			ctx := context.Background()
			type ctxKey struct{}
			newCtx := context.WithValue(ctx, ctxKey{}, 0)

			var got1, got2 context.Context
			var failed2, skipped2 bool
			var failed, skipped bool

			got := goyek.NewRunner(func(a *goyek.A) {
				a2 := a.WithContext(newCtx)
				got1 = a.Context()  // ctx
				got2 = a2.Context() // newCtx
				a2.Log("1")
				a.Cleanup(func() {
					skipped = a.Skipped()
					failed = a.Failed()
					a.Log("3")
				})
				a2.Cleanup(func() {
					skipped2 = a2.Skipped()
					failed2 = a2.Failed()
					a2.Log("2")
				})
				tc.fn(a, a2)
			})(goyek.Input{Context: ctx, Output: sb, Logger: goyek.FmtLogger{}})

			if got.Status != tc.wantStatus {
				t.Errorf("status was %s but want %s", got.Status, tc.wantStatus)
			}
			if got1 != ctx {
				t.Errorf("original Context returned %v but want %v", got1, ctx)
			}
			if got2 != newCtx {
				t.Errorf("derived Context returned %v but want %v", got2, newCtx)
			}
			if out, want := sb.String(), "1\n2\n3\n"; out != want {
				t.Errorf("logging or cleanup failed, out was %q but want %q", out, want)
			}
			if failed != tc.wantFailed {
				t.Errorf("original Failed returned %v but want %v", failed, tc.wantFailed)
			}
			if skipped != tc.wantSkipped {
				t.Errorf("original Skipped returned %v but want %v", skipped, tc.wantSkipped)
			}
			if failed2 != tc.wantFailed {
				t.Errorf("derived Failed returned %v but want %v", failed2, tc.wantFailed)
			}
			if skipped2 != tc.wantSkipped {
				t.Errorf("derived Skipped returned %v but want %v", skipped2, tc.wantSkipped)
			}
		})
	}
}
```

These tests failed:

```
--- FAIL: TestA_WithContext (0.00s)
    --- FAIL: TestA_WithContext/Pass (0.00s)
        a_test.go:200: logging or cleanup failed, out was "1\n3\n2\n" but want "1\n2\n3\n"
    --- FAIL: TestA_WithContext/Fail (0.00s)
        a_test.go:200: logging or cleanup failed, out was "1\n3\n2\n" but want "1\n2\n3\n"
        a_test.go:209: derived Failed returned false but want true
    --- FAIL: TestA_WithContext/FailDerived (0.00s)
        a_test.go:200: logging or cleanup failed, out was "1\n3\n2\n" but want "1\n2\n3\n"
    --- FAIL: TestA_WithContext/Skip (0.00s)
        a_test.go:200: logging or cleanup failed, out was "1\n3\n2\n" but want "1\n2\n3\n"
        a_test.go:212: derived Skipped returned false but want true
    --- FAIL: TestA_WithContext/SkipDerived (0.00s)
        a_test.go:200: logging or cleanup failed, out was "1\n3\n2\n" but want "1\n2\n3\n"
```

You can take a look at https://github.com/goyek/goyek/pull/460

File: a_test.go (hunk 15 above)
Comment (maniak89) [reply a_test.go:446]: I think it should be left in to make sure it doesn't cause any problems, there are no unnecessary tests at all. The main problem I had was with parallel=true, and I think it should be left in.

File: a_test.go (hunk 14 above)
Comment (maniak89) [reply a_test.go]: I agree - it seems redundant

File: a_test.go (hunk 16 above)
Comment (maniak89) [reply a_test.go:284]: Skip also

**Comment by maniak89** [comment]
> These tests failed

I will try to fix it. I need about 3 hours for my business and after that I will do it.

**Comment by maniak89** [comment]
Maybe withContext should have a signature that returns not only the new action, but also a cancel function that the user is required to call with defer?

**Comment by pellared** [comment]
> but also a cancel function that the user is required to call with defer?

This would result in a bad user experience, which indicated bad design.

**Comment by maniak89** [comment]
Absorbed https://github.com/goyek/goyek/pull/460
